File: data-structures/deque/Deque.java
public class Deque {
  Node front, rear;

  class Node {
    int data;
    Node next;
    Node prev;

    public Node(int data) {
      this.data = data;
      this.next = null;
      this.prev = null;
    }
  }

  public Deque() {
    front = rear = null;
  }

  public void addFront(int data) {
    Node newNode = new Node(data);
    if (isEmpty()) {
      front = rear = newNode;
    } else {
      newNode.next = front;
      front.prev = newNode;
      front = newNode;
    }
  }

  public void addRear(int data) {
    Node newNode = new Node(data);
    if (isEmpty()) {
      front = rear = newNode;
    } else {
      rear.next = newNode;
      newNode.prev = rear;
      rear = newNode;
    }
  }

  public void deleteFront() {
    if (isEmpty()) {
      System.out.println("Deque is empty");
      return;
    }

    if (front == rear) {
      front = rear = null;
    } else {
      front = front.next;
      front.prev = null;
    }
  }

  public void deleteLast() {
    if (isEmpty()) {
      System.out.println("Deque is empty");
      return;
    }

    if (front == rear) {
      front = rear = null;
    } else {
      rear = rear.prev;
      rear.next = null;
    }
  }

  public int peekFront() {
    if (isEmpty()) {
      throw new RuntimeException("Deque is empty");
    }
    return front.data;
  }

  public int peekRear() {
    if (isEmpty()) {
      throw new RuntimeException("Deque is empty");
    }
    return rear.data;
  }

  public void display() {
    if (isEmpty()) {
      System.out.println("Deque is empty");
      return;
    }

    Node current = front;
    System.out.print("Deque: ");
    while (current != null) {
      System.out.print(current.data);
      if (current.next != null) {
        System.out.print(" <-> ");
      }
      current = current.next;
    }
    System.out.println();
  }

  public boolean isEmpty() {
    return front == null;
  }
}

File: data-structures/deque/Deque.go
package main

type Node struct {
	data int
	prev *Node
	next *Node
}

type Deque struct {
	front *Node
	rear  *Node
}

func (dq *Deque) IsEmpty() bool {
	return dq.front == nil
}

func (dq *Deque) AddFront(data int) {
	newNode := &Node{data: data}
	if dq.IsEmpty() {
		dq.front = newNode
		dq.rear = newNode
	} else {
		newNode.next = dq.front
		dq.front.prev = newNode
		dq.front = newNode
	}
}

func (dq *Deque) AddRear(data int) {
	newNode := &Node{data: data}
	if dq.IsEmpty() {
		dq.front = newNode
		dq.rear = newNode
	} else {
		dq.rear.next = newNode
		newNode.prev = dq.rear
		dq.rear = newNode
	}
}

func (dq *Deque) PopFront() int {
	if dq.IsEmpty() {
		panic("Deque is empty")
	}
	data := dq.front.data
	if dq.front == dq.rear {
		dq.front = nil
		dq.rear = nil
	} else {
		dq.front = dq.front.next
		dq.front.prev = nil
	}
	return data
}

func (dq *Deque) PopRear() int {
	if dq.IsEmpty() {
		panic("Deque is empty")
	}
	data := dq.rear.data
	if dq.front == dq.rear {
		dq.front = nil
		dq.rear = nil
	} else {
		dq.rear = dq.rear.prev
		dq.rear.next = nil
	}
	return data
}

func (dq *Deque) PeekFront() int {
	if dq.IsEmpty() {
		panic("Deque is empty")
	}
	return dq.front.data
}

func (dq *Deque) PeekRear() int {
	if dq.IsEmpty() {
		panic("Deque is empty")
	}
	return dq.rear.data
}
